### PR TITLE
Better code around decompression [2]

### PR DESCRIPTION
**Pull request by nikitamikhaylov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

### CODE REVIEW

File: src/Compression/LZ4_decompress_faster.cpp [hunk 1]
```diff
@@ -439,11 +439,14 @@ bool NO_INLINE decompressImpl(
             {
                 s = *ip++;
                 length += s;
-            } while (unlikely(s == 255));
+            } while (ip < input_end && unlikely(s == 255));
```
Comment (KochetovNicolai) [inline src/Compression/LZ4_decompress_faster.cpp:442]: Maybe `while (unlikely(s == 255 && ip < input_end))`?

File: src/Compression/LZ4_decompress_faster.cpp [hunk 2]
```diff
@@ -464,18 +467,18 @@ bool NO_INLINE decompressImpl(
...
+        wildCopy<copy_amount>(op, ip, copy_end);    /// Here we can write up to copy_amount - 1 bytes after buffer.
+
+        if (copy_end == output_end)
+            return true;
 
         ip += length;
         op = copy_end;
 
-        if (unlikely(ip > input_end))
+        if (unlikely(ip + 1 >= input_end))
```
Comment (KochetovNicolai) [inline src/Compression/LZ4_decompress_faster.cpp:481]: Not clear to me why this change is better. Maybe comment?

**Review by KochetovNicolai — APPROVED** [review]
What is changed looks ok to me.
Not sure we didn't miss something

File: src/Compression/LZ4_decompress_faster.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Compression/LZ4_decompress_faster.cpp:481]: `!(ip + 1 < input_end) <=> (ip + 1) >= input_end)`

**Comment by Algunenano** [comment]
Are the [changes in performance](https://clickhouse-test-reports.s3.yandex.net/27743/2272c9896c7064cb7a6e3bc3da72c7d55ebe7c82/performance_comparison/report.html) unrelated?
